kinesis/src/main/scala/akka/stream/alpakka/kinesis/scaladsl/KinesisSchedulerSource.scala

```diff
@@ -0,0 +1,113 @@
+import scala.util.control.Exception
+import scala.util.{Failure, Success}
+
+object KinesisSchedulerSource {
+
+  def apply(
+      schedulerBuilder: ShardRecordProcessorFactory => Scheduler,
+      settings: KinesisSchedulerSourceSettings = KinesisSchedulerSourceSettings.defaultInstance
+  ): Source[CommittableRecord, Future[Scheduler]] =
+    Setup
+      .source { implicit mat => _ =>
+        import mat.executionContext
+        Source
+          .queue[CommittableRecord](settings.bufferSize, OverflowStrategy.backpressure)
+          .watchTermination()(Keep.both)
+          .mapMaterializedValue {
+            case (queue, watch) =>
+              val semaphore = new Semaphore(1, true)
+              val scheduler = schedulerBuilder(
+                new ShardRecordProcessorFactory {
+                  override def shardRecordProcessor(): ShardRecordProcessor =
+                    new ShardProcessor(
+                      record => {
+                        semaphore.acquire()
+                        (Exception.nonFatalCatch either Await
+                          .result(queue.offer(record), settings.backpressureTimeout) left)
+                          .foreach(err => queue.fail(BackpressureTimeout(err)))
+                        semaphore.release()
+                      }
+                    )
+                }
+              )
+
+              Future(scheduler.run()).onComplete {
+                case Failure(ex) =>
+                  queue.fail(SchedulerUnexpectedShutdown(ex))
+                case Success(_) => queue.complete()
+              }
+              watch.onComplete(_ => Future(scheduler.shutdown()))
+              scheduler
+          }
+      }
+
+  def checkpointRecordsFlow(
+      settings: KinesisSchedulerCheckpointSettings = KinesisSchedulerCheckpointSettings.defaultInstance
+  )(implicit schedulerExecutor: ExecutionContext): Flow[CommittableRecord, KinesisClientRecord, NotUsed] =
+    Flow[CommittableRecord]
+      .groupBy(MAX_KINESIS_SHARDS, _.shardId)
+      .groupedWithin(settings.maxBatchSize, settings.maxBatchWait)
+      .via(GraphDSL.create() { implicit b ⇒
+        import GraphDSL.Implicits._
+
+        val split =
+          b.add(scaladsl.Broadcast[immutable.Seq[CommittableRecord]](2))
+        val join = b.add(Zip[Done, immutable.Seq[CommittableRecord]])
+        val result = b.add(Flow[KinesisClientRecord])
+
+        split
+          .out(0)
+          .map(_.max)
+          .mapAsync(1)(
+            r =>
+              if (r.canBeCheckpointed()) r.tryToCheckpoint()
+              else Future.successful(Done)
+          ) ~> join.in0
+        split.out(1) ~> join.in1
+
+        join.out.map(_._2).mapConcat(identity).map(_.record) ~> result
+
+        FlowShape(split.in, result.out)
+      })
+      .mergeSubstreams
+      .withAttributes(ActorAttributes.supervisionStrategy {
+        case _: ShutdownException =>
+          Resume
+        case _ => Stop
+      })
```

Comment:
@ennru can you help me with this stage?

This stage is just a Flow that "checkpoints" Kinesis records in an optimized and shard-aware fashion. A "checkpoint" is actually a synchronous IO call, that's why I wrap it around a Future (`CommittableRecord.tryToCheckpoint()`).

Do you think that it makes sense to remove the EC parameter also from this stage? If yes, what approach would you take?

Reply:
An option is to internally keep a reference of the EC that comes from the `ActorMaterializer` in each `CommittableRecord` (what I did before the latest change, but instead of using user's EC, use the materializer one.

Would that be ok?

Reply:
Great that you found it, looks good.